File: lib/data/repositories/expense_repository.dart
import 'package:drift/drift.dart';

import '../local/database.dart';
import '../local/enums.dart';

/// A transaction joined with its (optional) category.
class TransactionWithCategory {
  final Transaction txn;
  final Category? category;
  const TransactionWithCategory(this.txn, this.category);
}

/// Aggregated total for a category over a period.
class CategoryTotal {
  final Category? category;
  final double total;
  const CategoryTotal(this.category, this.total);
}

class ExpenseRepository {
  ExpenseRepository(this._db);
  final AppDatabase _db;

  // ---- Categories -------------------------------------------------------
  Stream<List<Category>> watchCategories() {
    return (_db.select(_db.categories)
          ..orderBy([(c) => OrderingTerm(expression: c.name)]))
        .watch();
  }

  Future<List<Category>> allCategories() => _db.select(_db.categories).get();

  Future<int> addCategory(CategoriesCompanion c) =>
      _db.into(_db.categories).insert(c);

  // ---- Transactions -----------------------------------------------------
  Stream<List<TransactionWithCategory>> watchTransactions({
    DateTime? from,
    DateTime? to,
    int limit = 500,
  }) {
    final query = _db.select(_db.transactions).join([
      leftOuterJoin(
        _db.categories,
        _db.categories.id.equalsExp(_db.transactions.categoryId),
      ),
    ]);
    if (from != null) {
      query.where(_db.transactions.date.isBiggerOrEqualValue(from));
    }
    if (to != null) {
      query.where(_db.transactions.date.isSmallerThanValue(to));
    }
    query
      ..orderBy([OrderingTerm.desc(_db.transactions.date)])
      ..limit(limit);
    return query.watch().map((rows) {
      return rows
          .map((r) => TransactionWithCategory(
                r.readTable(_db.transactions),
                r.readTableOrNull(_db.categories),
              ))
          .toList();
    });
  }

  Future<int> upsertTransaction(TransactionsCompanion txn) {
    return _db.into(_db.transactions).insertOnConflictUpdate(txn);
  }

  Future<void> deleteTransaction(int id) {
    return (_db.delete(_db.transactions)..where((t) => t.id.equals(id))).go();
  }

  // ---- Aggregations -----------------------------------------------------
  /// Net total (income positive, expense negative) within range.
  Stream<double> watchNet(DateTime from, DateTime to) {
    return watchTransactions(from: from, to: to).map((rows) {
      var net = 0.0;
      for (final r in rows) {
        net += r.txn.type == TxnType.income ? r.txn.amount : -r.txn.amount;
      }
      return net;
    });
  }

  /// Sum of a single type within range.
  Future<double> sumByType(DateTime from, DateTime to, TxnType type) async {
    final amount = _db.transactions.amount.sum();
    final query = _db.selectOnly(_db.transactions)
      ..addColumns([amount])
      ..where(_db.transactions.date.isBiggerOrEqualValue(from) &
          _db.transactions.date.isSmallerThanValue(to) &
          _db.transactions.type.equalsValue(type));
    final row = await query.getSingle();
    return row.read(amount) ?? 0;
  }
}

File: web/src/db/expense-repository.ts
import { db } from './database';
import type {
  Category,
  Transaction,
  TransactionWithCategory,
} from '../types';
import { TxnType } from '../types';
import { useLiveQuery } from 'dexie-react-hooks';

export function useCategories() {
  return useLiveQuery(() => db.categories.orderBy('name').toArray(), []) ?? [];
}

export function useCategoriesByType(type: TxnType) {
  return useLiveQuery(() => db.categories.where('type').equals(type).toArray(), [type]) ?? [];
}

export async function addCategory(category: Omit<Category, 'id'>): Promise<number> {
  return db.categories.add(category as Category);
}

export function useTransactions(options?: { from?: Date; to?: Date; limit?: number }) {
  return useLiveQuery(async () => {
    let txns: Transaction[];

    if (options?.from && options?.to) {
      txns = await db.transactions
        .where('date')
        .between(options.from, options.to, true, true)
        .reverse()
        .toArray();
    } else {
      txns = await db.transactions.orderBy('date').reverse().toArray();
    }

    if (options?.limit) {
      txns = txns.slice(0, options.limit);
    }

    const categories = await db.categories.toArray();
    const catMap = new Map(categories.map((c: Category) => [c.id, c]));

    return txns.map((t: Transaction) => ({
      ...t,
      category: t.categoryId ? catMap.get(t.categoryId) : undefined,
    })) as TransactionWithCategory[];
  }, [options?.from?.getTime(), options?.to?.getTime(), options?.limit]) ?? [];
}

export async function upsertTransaction(txn: Omit<Transaction, 'id'> & { id?: number }): Promise<number> {
  if (txn.id) {
    await db.transactions.update(txn.id, txn);
    return txn.id;
  }
  return db.transactions.add(txn as Transaction);
}

export async function deleteTransaction(id: number): Promise<void> {
  await db.transactions.delete(id);
}

export function useNet(from: Date, to: Date) {
  return useLiveQuery(async () => {
    const txns = await db.transactions
      .where('date')
      .between(from, to, true, true)
      .toArray();

    return txns.reduce((sum: number, t: Transaction) => {
      return sum + (t.type === TxnType.Income ? t.amount : -t.amount);
    }, 0);
  }, [from.getTime(), to.getTime()]) ?? 0;
}

export async function sumByType(from: Date, to: Date, type: TxnType): Promise<number> {
  const txns = await db.transactions
    .where('date')
    .between(from, to, true, true)
    .toArray();

  return txns
    .filter((t: Transaction) => t.type === type)
    .reduce((sum: number, t: Transaction) => sum + t.amount, 0);
}
